doc: various explainer updates (#40)
tc39/proposal-module-attributes: README.md

Some developers have the intuition that the file extension could be used to determine the module type, as it is in many existing non-standard module systems. However, it's a deep web architectural principle that the suffix of the URL (which you might think of as the "file extension" outside of the web) does not lead to semantics of how the page is interpreted. In practice, on the web, there is a widespread [mismatch between file extension and the HTTP Content Type header](content-type-vs-file-extension.md). All of this sums up to it being infeasible to depend on file extensions/suffixes included in the module specifier to be the basis for this checking.

> There are other possible pieces of metadata which could be associated with modules, see [#8](https://github.com/littledan/proposal-module-attributes/issues/8) for further discussion.

Reviewer: ```suggestion
There are other possible pieces of metadata which could be associated with modules, see [#8](https://github.com/tc39/proposal-module-attributes/issues/8) for further discussion.
```

This proposal is at Stage 1.

> Standardization here would consist of building consensus not just in TC39 but also in WHATWG HTML as well as the Node.js ESM effort and a general audit of semantic requirements across various host environments ([#10](https://github.com/littledan/proposal-module-attributes/issues/10), [#24](https://github.com/littledan/proposal-module-attributes/issues/24) and [#25](https://github.com/littledan/proposal-module-attributes/issues/25)).

Reviewer: ```suggestion
Standardization here would consist of building consensus not just in TC39 but also in WHATWG HTML as well as the Node.js ESM effort and a general audit of semantic requirements across various host environments ([#10](https://github.com/tc39/proposal-module-attributes/issues/10), [#24](https://github.com/tc39/proposal-module-attributes/issues/24) and [#25](https://github.com/tc39/proposal-module-attributes/issues/25)).
```

import value from "module" as {key1: "value1", key2: [1, 2, 3]};
```

This would allow module attributes to scale to support a variety of metadata. It is currently unclear if the generalized constant form will be pursued in this proposal or in a follow up. In either case the addition of the generalized constant form is not hard to design, specify and implement on top of the existing [spec text outline](https://tc39.github.io/proposal-module-attributes/). 

## FAQ

### Why not out of band?

Why not both? The champions of this proposal think that exploring both an in- and out of band solutions to this problem are desirable. Depending on the environment in which JavaScript is being executed in either approach could be seen as desirable.

> While an in band solution is more verbose, it is also more straightforward for developers to adopt. For smaller projects developers do not need to create an extra file by hand. For large project with many dependencies developers will not have to worry about creating a large manifest by compiling the metadata of all of their dependencies. Module authors will also not have to worry about shipping a manifest in order for consumers to be able to run their modules.

Reviewer 2: It may be worth noting that only native web module authors would have this worry; anyone who ships a node module already has that information contained in the file extension, for a top-level build tool to easily pick up.

Author: That isn't neccesarily true. Deno, another serverside environment, utilizes https URLs ... and there is a possible future where node could support https URLs

> The [import-maps proposal](https://github.com/WICG/import-maps) is a great example of an out of band manifest that could be created at installation time, but its focus is on modules and at this time all necessary information to generate an application's import map could be found in the package.json of its dependencies.

Reviewer 3: I'm confused by this paragraph. How does this relate to the rest?

Nit: s/its/it's/

Author: dropped paragraph

### Are there cross-environment concerns?

> Absolutely. This proposal is not attempting to standardize the specific `type`s a host will be allowed to support. It will be up to each runtime to decide which `type`s to support, as well as how to handle the the case of an unsupported module `type` being imported.

Reviewer 3: s/runtime/host enviroment/

Module specifiers are an example of prior art for this type of decision making. It is left up to the host how to resolve a module specifier into a URL that will eventually be loaded.

Independent of the potential cross-environment concerns it is not a problem that is specific to an in band solution. An out of band solution would also suffer from the risk of inconsistent implementation or support across runtimes.

The topic of attribute divergence is further discussed in  [#34](https://github.com/tc39/proposal-module-attributes/issues/34).

### Should the generalized constant form be supported in the first iteration?

> This proposal was initially created to serve a very specific purpose, unblocking security concerns raised about importing alternative module types (JSON, CSS, HTML). While there is no debate that the generalized constant form would be useful, it does pose the risk of increasing cross-environment concerns mentioned above.

Reviewer 3: New wording:
"This proposal was inspired by a single use case, to unblock other module types (JSON, CSS, HTML) from a security concern on environments like the Web where modules may be remotely loaded. It's unclear whether there is a sufficiently motivating use case to add the generalized constant form in the first iteration of the proposal. The generalized constant form adds some complexity, including in the surface area of cross-environment concerns."

> The champion group is very much in support of pursuing the generalized constant form in the long term, but would like to discuss with committee if it should be supported in the first iteration or not.

Reviewer 3: New wording:
"The champion group is interested in investigating the generalized constant form, and considering it for inclusion in either this proposal or a follow-on, weighing the expressiveness against the costs listed above, in consultation with TC39."

### How would this proposal work with caching?

Currently in [the ECMA262 spec](https://tc39.es/ecma262/#sec-hostresolveimportedmodule) passing the same module specifier and referrer there is guaranteed to get the same module. This proposal aims to have a similar guarantee; if a developer passes the same module specifier, referrer, and attributes there would be a guarantee of receiving the same module.

> It is possible that hosts will explore additional caching and coalescing beyond this.

Reviewer 3: New wording:
"We expect that hosts will cache/coalesce further than this. For example, on the Web and similar environments, when providing a type module attribute, the module type would not form part of the cache key, where modules are always cached by the resolved URL. Instead, the type attribute is used simply for a check, which would cause the module graph to fail to load in the event of a mismatch."

import value from "module" as {key1: "value1", key2: [1, 2, 3]};
```

> This would allow module attributes to scale to support a variety of metadata. It is currently unclear if the generalized constant form will be pursued in this proposal or in a follow up. In either case the addition of the generalized constant form is not hard to implement on top of the existing [spec text](https://tc39.github.io/proposal-module-attributes/). 

Reviewer 3: s/implement/design, specify and implement/

Reviewer 3: "existing spec text outline" (the spec text is not done)

In TC39 conventions, "implementation" happens in languages like C++; we don't tend to use that word for specification work

Author: fixed

## FAQ

### Why not out of band?

Why not both? The champions of this proposal think that exploring both an in- and out of band solutions to this problem are desirable. Depending on the environment in which JavaScript is being executed in either approach could be seen as desirable.

> While an in band solution is more verbose, it is also more straightforward for developers to adopt. For large project with many dependencies developers will not have to worry about creating a large manifest by compiling the meta data of all of their dependencies. Module authors will also not have to worry about shipping manifest in order for consumers to be able to run their modules.

Reviewer 2: ```suggestion
While an in band solution is more verbose, it is also more straightforward for developers to adopt. For large project with many dependencies developers will not have to worry about creating a large manifest by compiling the metadata of all of their dependencies. Module authors will also not have to worry about shipping a manifest in order for consumers to be able to run their modules.
```

> While an in band solution is more verbose, it is also more straightforward for developers to adopt. For large project with many dependencies developers will not have to worry about creating a large manifest by compiling the meta data of all of their dependencies. Module authors will also not have to worry about shipping manifest in order for consumers to be able to run their modules.

Reviewer 3: Could we also mention that it's less verbose for smaller projects developed in a simple way, because they can just write the metadata inline instead of having to create this extra manifest by hand?

Author: done

> The [import-maps proposal][] is a great example of an out of band manifest that could be created at installation time, but it's focus is on modules and at this time all necessary information to generate an applications import map could be found in the package.json of it's dependencies.

Reviewer 2: ```suggestion
The [import-maps proposal][] is a great example of an out of band manifest that could be created at installation time, but its focus is on modules and at this time all necessary information to generate an application's import map could be found in the package.json of its dependencies.
```

> While there very well could be a future where an out of band manifest could be used for the purpose of identifying module type, the amount of infrastructure necessary for ecosystem adoption is significantly higher than the in band solution proposed here.

Reviewer 2: given that the infrastructure required for in-band includes all the change any new syntax requires - eslint, babel, typescript, editors, browsers/engines - while out-of-band does not, can you elaborate a bit on how you're measuring "higher"?

Reviewer 3: I think we could say things like, that identifying where to find the manifest is an extra piece of inherent complexity and ecosystem divergence: On the web, it may be a script tag; in Node.js, it may be package.json. Inline metadata eliminates this complexity in a comprehensive way. This forms an ecosystem adoption barrier; maybe that's what Myles is getting at.

I'd agree with @ljharb that "it's too much work to design this" is not a very strong argument (since we should probably do this anyway, e.g., for integrity) but I doubt that's what @MylesBorins was saying here.

Author: I've dropped the paragraph about infrastrucutre. my intent was that an out of band solution will require new forms of infrastructure and tooling that doesn't already exist. While new syntax does require updates to all of the tools you mentioned that is not particularly different than any other new syntax we add to the language. It also is an update that can happen under developers without requiring them to adopt any new tooling.

I'll think a bit more about how to best present this point, but for now I've dropped it as it seemed to not be having the desired impact

### Are there cross-environment concerns?

> Absolutely. This proposal is not attempting to standardize what `types` hosts will be allowed to support. It will be up to each runtime to decide which `types` to support, as well as how to handle the the case of an unsupported module `type` being imported.

Reviewer 3: Nit: Probably the final `s` should be outside of the backtick.

Author: Fixed

> Module specifiers are an example of prior art for this type of decision making. It is left up to the host how to resolve a module specifier into a URL that will eventually be loaded. The [import-maps proposal][] is a great example of work done outside of TC39 to allow hosts to share behavior for undefined parts of the spec.

Reviewer 3: I find citing import-maps here a bit odd. import-maps's specification is about as cross-environment as HTML's module loading sections, and both have been implemented cross-environment anyway.

Author: Just dropped the reference for now, we can revisit

Independent of the potential cross-environment concerns it is not a problem that is specific to an in band solution. An out of band solution would also suffer from the risk of inconsistent implementation or support across runtimes.

> 

Reviewer: Could we mention https://github.com/tc39/proposal-module-attributes/issues/34?

Author: Done

### Why not pursue the generalized constant form today?

This proposal was initially created to serve a very specific purpose, unblocking security concerns raised about importing alternative module types (JSON, CSS, HTML). While there is no debate that the generalized constant form would be useful, it does pose the risk of increasing cross-environment concerns mentioned above.

> The champion group is very much in support of pursuing the generalized constant form, but think that it would be prudent to focus the proposal to a single string initially.

Reviewer 3: I think we should be open to supporting the generalized constant form in the initial version.

Author: Took a stab at improving the language PTAL

### How would this proposal work with caching?

> At the moment there is nothing in the specification text that refers to caching. While a caching strategy would likely be dependent on the type of metadata included, for the generalized constant form, this proposal would defer to the policy of environments to decide how and what to cache.

Reviewer 3: This isn't true. See https://tc39.es/ecma262/#sec-hostresolveimportedmodule, the third bullet point. The specification requires that, if you pass up the same module specifier and referrer, you get the same module. In practice, there's *additional* caching that's shared between multiple referrers.

After thinking about this some more, I think we should use a similar guarantee when module attributes are included: If you pass the same specifier, referrer *and module attributes*, you get the same module. But hosts will probably do additional caching and coalescing beyond that. (E.g., in HTML, with types, it would always give you the same module based on the same specifier and referrer, or an error.)

Author: Took a stab at text PTAL

## Specification

> * [Ecmarkup source](./spec.html)

Reviewer 3: I don't think it's useful to link to the Ecmarkup source (even if many repos do that)

> * [Ecmarkup source](./spec.html)

Reviewer: ```suggestion
```

I agree with @littledan

> * [HTML Version](https://tc39.github.io/proposal-module-attributes/)

Reviewer 3: Please identify this as a "specification outline"--it's not filled in yet. And it's nice to use the tc39.es URLs.

Author: Fixed

### Should the generalized constant form be supported in the first iteration?

This proposal was initially created to serve a very specific purpose, unblocking security concerns raised about importing alternative module types (JSON, CSS, HTML). While there is no debate that the generalized constant form would be useful, it does pose the risk of increasing cross-environment concerns mentioned above.

The champion group is very much in support of pursuing the generalized constant form in the long term, but would like to discuss with committee if it should be supported in the first iteration or not.

### How would this proposal work with caching?

Currently in [the ECMA262 spec](https://tc39.es/ecma262/#sec-hostresolveimportedmodule) passing the same module specifier and referrer there is guaranteed to get the same module. This proposal aims to have a similar guarantee; if a developer passes the same module specifier, referrer, and attributes there would be a guarantee of receiving the same module.

It is possible that hosts will explore additional caching and coalescing beyond this.

## Specification

* [Specification Outline](https://tc39.es/proposal-module-attributes/)

> 

Reviewer: ```suggestion
```

> [import-maps proposal]: https://github.com/WICG/import-maps

Reviewer: ```suggestion
* [import-maps proposal]: https://github.com/WICG/import-maps
```
